operations/rack_resiliency/rack_resiliency.md

```diff
@@ -0,0 +1,490 @@
+
+The placement validation algorithm, as shown in the flow chart above decides whether the current placement of management nodes is suitable for enabling rack resiliency. If it is found that the current placement is not suitable for rack resiliency the next step which involves zoning for `Kubernetes` and `Ceph` is not performed.
+
+This module also evaluates if managed nodes are present in the management rack and informational messages are generated for the same.
+
+Note: Slingshot switch placement discovery and validation is out of scope for this release (CSM 1.7.0) of Rack Resiliency solution.
+
+### 2 Spreading of NCNs across `Kubernetes` zones, and spreading storage nodes across `Ceph` zones
+
+Upon successful validation of placement, zoning is performed.
+
+#### 2.1 Zoning `Kubernetes` nodes 
+
+The `Kubernetes` topology zoning can be used to apply labels to the racks/ nodes in order to create management failure domains (MFD).
+
+The node is labeled with the key `topology.kubernetes.io/zone=<zone-id>`, where `<zone-id>` is of the form `Rack-1`, `Rack-2`, and so on. These labels can be used to identify all the management nodes which belong to the same zone and is used to schedule the critical services across the zones. 
+
+
+#### 2.2 Zoning `Ceph`
+
+`Ceph` provides the CRUSH map algorithm which can help segregate the storage nodes across zones. Using a combination of CRUSH rules and bucket types (hosts, racks, rows, etc.), the data can be replicated across zones. The objective of `Ceph` zoning is to make sure data gets replicated at rack level, so that there would not be data loss in case of a rack failure.
+
+#### 2.2.1 Creating Ceph zones with CRUSH 
+
+* Create racks by creating `Ceph` buckets and map storage nodes to these buckets using 
+    * "`Ceph` osd crush add-bucket <rack_name> rack"
+    * "`Ceph` osd crush move {bucket-name} {bucket-type}={bucket-name}, [...]"
+* Create a CRUSH rule with rack as a failure domain
+* Map above rule to replicated `Ceph` pools
+
+#### 2.2.2 `Ceph` service zoning
+
+Along with `Ceph` data zoning, `Ceph` services also need to be zoned. These services include:
+
+* `Ceph` monitor
+* manager
+* mds service for admin-tools
+* cephfs
+
+### 3 Distributing critical services across Kubernetes zones
+
+One of the key factors which can support that CSM critical services will survive the failure of nodes or a single rack is to spread the replicas of these services across multiple zones spread across racks.
+
+Some of the services in CSM have already established pod affinities to spread the replicas across nodes. Since the nodes which are picked by `Kubernetes` scheduler can be on the same rack, it becomes necessary to include a topology constraint for these services which can help `Kubernetes` scheduler distribute the replicas across zones. This is achieved using the "Pod Topology Spread Constraints" feature and implemented through a policy in `kyverno`.

```

Comment:
```suggestion
Some of the services in CSM have already established pod affinities to spread the replicas across nodes. Because the nodes which are picked by the Kubernetes scheduler can be on the same rack, it is necessary to include a topology constraint for these services; this helps the Kubernetes scheduler distribute the replicas across zones. This is achieved using the "Pod Topology Spread Constraints" feature and is implemented through a policy in [Kyverno](../kubernetes/Kyverno.md).
```